File: JavaScript/1930-Unique Length-3 Palindromic Subsequences.js
/**
 * @param {string} s
 * @return {number}
 */
var countPalindromicSubsequence = function(s) {
    let leftIndex = new Map(), rightIndex = new Map();
    
    for (let i=0; i<s.length; i++){
        if (!leftIndex.has(s[i])){
            leftIndex.set(s[i], i);
        }
        rightIndex.set(s[i], i);
    }
    
    let res = 0;
    let visited = new Set();
    for (let key of leftIndex.keys()){
        for (let i=leftIndex.get(key)+1; i<rightIndex.get(key); i++){
            if (!visited.has(`${key}-${s[i]}`)){
                visited.add(`${key}-${s[i]}`);
                res++;
            }
        }
    }
    return res;
};

File: C++/1930-Unique Length-3 Palindromic Subsequences.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        vector<int> cnt(26);
        for (char c:s) cnt[c-'a']++;
        vector<int> pre(26);
        vector<vector<int>> visited(26, vector<int>(26, 0));
        int res = 0;

        for (char c:s) {
            cnt[c-'a']--;
            for (int i=0; i<26; i++) {
                if (pre[i] > 0 && cnt[i] > 0 && !visited[c-'a'][i]) {
                    visited[c-'a'][i] = 1;
                    res++;
                }
            }
            pre[c-'a']++;
        }

        return res;
    }
};